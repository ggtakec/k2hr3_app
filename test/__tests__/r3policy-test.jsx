/*
 *
 * K2HR3 Web Application
 *
 * Copyright 2017 Yahoo! Japan Corporation.
 *
 * K2HR3 is K2hdkc based Resource and Roles and policy Rules, gathers
 * common management information for the cloud.
 * K2HR3 can dynamically manage information as "who", "what", "operate".
 * These are stored as roles, resources, policies in K2hdkc, and the
 * client system can dynamically read and modify these information.
 *
 * For the full copyright and license information, please view
 * the license file that was distributed with this source code.
 *
 * AUTHOR:   Takeshi Nakatani
 * CREATE:   Mon Jul 10 2018
 * REVISION:
 *
 */

import React					from 'react';										// eslint-disable-line no-unused-vars
import renderer					from 'react-test-renderer';
import getElementWithContext	from 'react-test-context-provider';					// for context provider
import { ThemeProvider }		from '@material-ui/styles';							// for custom theme
import CssBaseline				from '@material-ui/core/CssBaseline';				// for reset.css

import r3Theme					from '../../src/components/r3theme';				// custom theme
import R3Policy					from '../../src/components/r3policy';
import R3Provider				from '../../src/util/r3provider';

import mock_fetch				from '../__mocks__/fetchMock';						// eslint-disable-line no-unused-vars
import { createNodeMock }		from '../__mocks__/materialUiMock';					// for material-ui

//
// Mock functions
//
// [NOTE]
// If you need to customize return value, you can call methods
// for return value by each mock function.
// see: https://jestjs.io/docs/ja/mock-functions
//
// ex)	mockfunc
//			.mockReturnValueOnce(10)
//			.mockReturnValueOnce('x')
//			.mockReturnValue(true);
//
const save		= jest.fn();											// eslint-disable-line no-undef
const update	= jest.fn();											// eslint-disable-line no-undef

//
// Mock functions for parts in TextField
//
// [NOTE]
// Material-ui's TextField has some components these set random "id" to itself.
// The random id is created by Math functions.
//		ex: "Math.floor(Math.random() * 0xFFFF)"
// So we need fixed id to compare snapshots, set following mock function.
//
Math.random = jest.fn(() => 0);											// eslint-disable-line no-undef

//
// Dummy datas
//
const policy = {
	action:		[ 'yrn:yahoo::::action:read' ],
	alias:		[],
	condition:	[],
	effect:		'allow',
	reference:	0,
	resource:	[ 'yrn:yahoo:::33348:resource:k2hr3_entest_obj_res_01' ]
};

//
// Main test
//
describe('R3Policy', () => {											// eslint-disable-line no-undef
	it('test snapshot for R3Policy', () => {							// eslint-disable-line no-undef
		// [NOTE]
		// R3Policy has autoWidth property which is set to DropDownMenu component through R3Policy
		// The reason why we defined autoWidth as a property of R3Policy is that JEST+material-ui
		// test detects the following errors.
		//		TypeError: Can not read property 'style' of undefined
		// This error occurred in DropDownMenu.js file. Property of DropDownMenu autoWidth works
		// as true if undefined, but it must be false when using JEST. For that, autoWidth is
		// defined in the property of R3Policy, and in this test it is set to false.
		//

		/* eslint-disable indent */
		const r3provider	= new R3Provider(null);
		const element		= getElementWithContext({
									r3Context:	r3provider.getR3Context()
								},
								<ThemeProvider theme={ r3Theme } >
									<CssBaseline />
									<R3Policy
										r3provider={ r3provider }
										policy={ policy }
										dispUnique={ 1 }
										onSave={ save }
										onUpdate={ update }
										isReadMode={ false }
										autoWidth={ false }
									/>
								</ThemeProvider>
							);
		/* eslint-enable indent */

		const component = renderer.create(element, { createNodeMock });
		let tree		= component.toJSON();
		expect(tree).toMatchSnapshot();									// eslint-disable-line no-undef
	});
});

/*
 * VIM modelines
 *
 * vim:set ts=4 fenc=utf-8:
 *
 */
