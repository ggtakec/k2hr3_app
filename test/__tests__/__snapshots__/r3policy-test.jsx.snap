// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`R3Policy test snapshot for R3Policy 1`] = `
<div
  className="R3Policy-root-1"
>
  <h6
    className="MuiTypography-root R3Policy-subTitleTop-2 MuiTypography-subtitle2 MuiTypography-colorPrimary MuiTypography-noWrap"
  >
    EFFECT
  </h6>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline R3Policy-effectSelect-4"
    onClick={[Function]}
  >
    <div
      aria-haspopup="listbox"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      ALLOW
    </div>
    <input
      onAnimationStart={[Function]}
      type="hidden"
      value="allow"
    />
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
  <h6
    className="MuiTypography-root R3Policy-subTitle-3 MuiTypography-subtitle2 MuiTypography-colorPrimary MuiTypography-noWrap"
  >
    ACTION
  </h6>
  <div
    className="MuiFormGroup-root MuiFormGroup-row"
  >
    <label
      className="MuiFormControlLabel-root R3Policy-actionLeftLabel-6"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-119 MuiCheckbox-root MuiCheckbox-colorSecondary R3Policy-actionCheckbox-5 PrivateSwitchBase-checked-120 Mui-checked MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={true}
            className="PrivateSwitchBase-input-122"
            data-indeterminate={false}
            disabled={false}
            onChange={[Function]}
            type="checkbox"
            value="yrn:yahoo::::action:read"
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </span>
      <span
        className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-noWrap"
        >
          READ
        </p>
      </span>
    </label>
    <label
      className="MuiFormControlLabel-root R3Policy-actionEndLabel-8"
    >
      <span
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-119 MuiCheckbox-root MuiCheckbox-colorSecondary R3Policy-actionCheckbox-5 MuiIconButton-colorSecondary"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <input
            checked={false}
            className="PrivateSwitchBase-input-122"
            data-indeterminate={false}
            disabled={false}
            onChange={[Function]}
            type="checkbox"
            value="yrn:yahoo::::action:write"
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </span>
      <span
        className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-noWrap"
        >
          WRITE
        </p>
      </span>
    </label>
  </div>
  <h6
    className="MuiTypography-root R3Policy-subTitle-3 MuiTypography-subtitle2 MuiTypography-colorPrimary MuiTypography-noWrap"
  >
    RESOURCE(YRN)
  </h6>
  <div
    className="R3Policy-enclosureElement-20"
  >
    <div
      className="MuiFormControl-root MuiTextField-root R3Policy-resourceTextField-9"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline R3Policy-inputTextField-19 MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="policyResource_0"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Input resource YRN path"
          required={false}
          rows={1}
          type="text"
          value="yrn:yahoo:::33348:resource:k2hr3_entest_obj_res_01"
        />
      </div>
    </div>
    <button
      aria-describedby={null}
      aria-label="delete resource"
      className="MuiButtonBase-root MuiIconButton-root R3Policy-deleteResourceButton-11"
      disabled={false}
      label="delete-resource"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      title="Delete resource YRN path"
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.3 5.71a.9959.9959 0 00-1.41 0L12 10.59 7.11 5.7a.9959.9959 0 00-1.41 0c-.39.39-.39 1.02 0 1.41L10.59 12 5.7 16.89c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L12 13.41l4.89 4.89c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    className="R3Policy-enclosureElement-20"
  >
    <div
      className="MuiFormControl-root MuiTextField-root R3Policy-resourceTextField-9"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline R3Policy-inputTextField-19 MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="resourceNew"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Input resource YRN path"
          required={false}
          rows={1}
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      aria-describedby={null}
      aria-label="add resource"
      className="MuiButtonBase-root MuiIconButton-root R3Policy-addResourceButton-10"
      disabled={false}
      label="add-resource"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      title="Add resource YRN path"
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
  <h6
    className="MuiTypography-root R3Policy-subTitle-3 MuiTypography-subtitle2 MuiTypography-colorPrimary MuiTypography-noWrap"
  >
    ALIASES
  </h6>
  <div
    className="R3Policy-enclosureElement-20"
  >
    <div
      className="MuiFormControl-root MuiTextField-root R3Policy-aliasTextField-13"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline R3Policy-inputTextField-19 MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name="aliasNew"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Input alias YRN path"
          required={false}
          rows={1}
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      aria-describedby={null}
      aria-label="add alias"
      className="MuiButtonBase-root MuiIconButton-root R3Policy-addAliasButton-16"
      disabled={false}
      label="add-alias"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      title="Add alias"
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    className="R3FormButtons-root-148"
  >
    <button
      aria-label="cancel"
      className="MuiButtonBase-root MuiButton-root MuiButton-contained R3FormButtons-cancelButton-149 MuiButton-containedPrimary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        CANCEL
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root R3FormButtons-buttonIcon-151"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
          />
        </svg>
      </span>
    </button>
    <button
      aria-label="save"
      className="MuiButtonBase-root MuiButton-root MuiButton-contained R3FormButtons-saveButton-150 MuiButton-containedSecondary Mui-disabled Mui-disabled"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={-1}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        SAVE
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root R3FormButtons-buttonIcon-151"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
